tests/unit/OccAnalysisSpec.hs

```diff
@@ -0,0 +1,339 @@
+  -- or just initTopState, to always compile the prelude during unit tests?
+  init_env <- runIO loadCache
+  (_, env) <- runIO $ runTopperM cfg init_env $ ensureModuleLoaded Prelude
+  describe "Occurrence analysis" do
+    it "counts a reference as a use" do
+      ann <- analyze cfg env
+        [ ":p"
+        , "  xs : (Fin 10) => Float = unreachable ()"
+        , "  xs"
+        ]
+      ann `shouldBe` OccInfo (UsageInfo 1 (0,Bounded 1))
+    it "counts indexing in a for as one use" do
+      ann <- analyze cfg env
+        [ ":p"
+        , "  xs : (Fin 10) => Float = unreachable ()"
+        , "  for i. xs.i"
+        ]
+      ann `shouldBe` OccInfo (UsageInfo 1 (1,Bounded 1))
+    it "counts indexing depth in nested fors" do
+      ann <- analyze cfg env
+        [ ":p"
+        , "  xs : (Fin 10) => (Fin 3) => Float = unreachable ()"
+        , "  for i j. xs.i.j"
+        ]
+      ann `shouldBe` OccInfo (UsageInfo 1 (2,Bounded 1))
+    it "counts two array uses" do
+      ann <- analyze cfg env
+        [ ":p"
+        , "  xs : (Fin 10) => Float = unreachable ()"
+        , "  (for i. xs.i, for j. xs.j)"
+        ]
+      ann `shouldBe` OccInfo (UsageInfo 2 (1,Bounded 2))
+    it "counts array and non-array uses" do
+      ann <- analyze cfg env
+        [ ":p"
+        , "  xs : (Fin 10) => Float = unreachable ()"
+        , "  (for i. xs.i, xs)"
+        ]
+      ann `shouldBe` OccInfo (UsageInfo 2 (1,Bounded 2))
+    it "counts different case arms as static but not dynamic uses" do
+      ann <- analyze cfg env
+        [ ":p"
+        , "  xs : (Fin 10) => Float = unreachable ()"
+        , "  if unreachable ()"
+        , "    then for i. xs.i"
+        , "    else for j. xs.j"
+        ]
+      ann `shouldBe` OccInfo (UsageInfo 2 (1,Bounded 1))
+    it "understands one index injection" do
+      ann <- analyze cfg env
+        [ ":p"
+        , "  xs : ((Fin 10) | (Fin 4)) => Float = unreachable ()"
+        , "  for j. xs.(Left j)"
+        ]
+      ann `shouldBe` OccInfo (UsageInfo 1 (1,Bounded 1))
+    it "understands distinct index injections" do
+      ann <- analyze cfg env
+        [ ":p"
+        , "  xs : ((Fin 10) | (Fin 4)) => Float = unreachable ()"
+        , "  (for i. xs.(Left i), for j. xs.(Right j))"
+        ]
+      ann `shouldBe` OccInfo (UsageInfo 2 (1,Bounded 1))
+    it "detects and eschews index arithmetic" do
+      ann <- analyze cfg env
+        [ ":p"
+        , "  xs : (Fin 4) => Float = unreachable ()"
+        , "  for i:(Fin 3). xs.((ordinal i + 1)@_)"
+        ]
+      ann `shouldBe` OccInfo (UsageInfo 1 (1,Unbounded))
+    it "detects non-nested single-uses cases despite index arithmetic" do
+      ann <- analyze cfg env
+        [ ":p"
+        , "  xs : (Fin 4) => Float = unreachable ()"
+        , "  xs.(1@_)"
+        ]
+      -- Arguably, should be able to prove that zero levels of exposed indexing
+      -- (not one) suffice for inlining xs to be safe here, but doesn't prove it
```

Comment:
How come "zero levels"? xs is indexed, right?

Reply:
Yes, but inlining `xs` is work-preserving here even if we can't cancel against the indexing operation, so the most precise answer is zero levels.